server/src/main/java/org/opensearch/common/time/RFC3339DateTimeFormatter.java

```diff
@@ -0,0 +1,422 @@
+ *    Complete date plus hours and minutes:
+ *       YYYY-MM-DDThh:mmTZD (eg 1997-07-16T19:20+01:00)
+ *    Complete date plus hours, minutes and seconds:
+ *       YYYY-MM-DDThh:mm:ssTZD (eg 1997-07-16T19:20:30+01:00)
+ *    Complete date plus hours, minutes, seconds and a decimal fraction of a second
+ *       YYYY-MM-DDThh:mm:ss.sTZD (eg 1997-07-16T19:20:30.45+01:00)
+ *       YYYY-MM-DDThh:mm:ss,sTZD (eg 1997-07-16T19:20:30,45+01:00)
+ * where:
+ *
+ *      YYYY = four-digit year
+ *      MM   = two-digit month (01=January, etc.)
+ *      DD   = two-digit day of month (01 through 31)
+ *      hh   = two digits of hour (00 through 23) (am/pm NOT allowed)
+ *      mm   = two digits of minute (00 through 59)
+ *      ss   = two digits of second (00 through 59)
+ *      s    = one or more(max 9) digits representing a decimal fraction of a second
+ *      TZD  = time zone designator (Z or z or +hh:mm or -hh:mm)
+ * </pre>
+ */
+final class RFC3339DateTimeFormatter extends OpenSearchDateTimeFormatter {
+    public static final char DATE_SEPARATOR = '-';
+    public static final char TIME_SEPARATOR = ':';
+    public static final char SEPARATOR_UPPER = 'T';
+    private static final char PLUS = '+';
+    private static final char MINUS = '-';
+    private static final char SEPARATOR_LOWER = 't';
+    private static final char SEPARATOR_SPACE = ' ';
+    private static final char FRACTION_SEPARATOR_1 = '.';
+    private static final char FRACTION_SEPARATOR_2 = ',';
+    private static final char ZULU_UPPER = 'Z';
+    private static final char ZULU_LOWER = 'z';
+
+    private ZoneId zone;
+
+    public RFC3339DateTimeFormatter(String pattern) {
+        super(pattern);
+    }
+
+    public RFC3339DateTimeFormatter(java.time.format.DateTimeFormatter formatter) {
+        super(formatter);
+    }
+
+    public RFC3339DateTimeFormatter(java.time.format.DateTimeFormatter formatter, ZoneId zone) {
+        super(formatter);
+        this.zone = zone;
+    }
+
+    @Override
+    public OpenSearchDateTimeFormatter withZone(ZoneId zoneId) {
+        return new RFC3339DateTimeFormatter(getFormatter().withZone(zoneId), zoneId);
+    }
+
+    @Override
+    public OpenSearchDateTimeFormatter withLocale(Locale locale) {
+        return new RFC3339DateTimeFormatter(getFormatter().withLocale(locale));
+    }
+
+    @Override
+    public Object parseObject(String text, ParsePosition pos) {
+        try {
+            return parse(text);
+        } catch (DateTimeException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public TemporalAccessor parse(final String dateTime) {
+        OffsetDateTime parsedDatetime = parse(dateTime, new ParsePosition(0));
+        return zone == null ? parsedDatetime : parsedDatetime.atZoneSameInstant(zone);
+    }
+
+    public OffsetDateTime parse(String date, ParsePosition pos) {
+        if (date == null) {
+            throw new NullPointerException("date cannot be null");
```

Comment:
Its a very debatable topic b/w IA and NP exception
https://stackoverflow.com/questions/3881/illegalargumentexception-or-nullpointerexception-for-a-null-parameter

I tend to favour NP exception just because its more specific in nature. Wdyt?

Reply:
He throws an NPE because that is what ITU does which does that because that is what java.time.OffsetDateTime does